### PR TITLE
[influxdb] Add option for using metadata value as measurement name

### PR DESCRIPTION
**Pull request by DerOetzi**
Add option for using the metadata value as measurement name. Part of solution for #9556 

Signed-off-by: Johannes Ott <info@johannes-ott.net>

### CODE REVIEW

**Review by lujop — COMMENTED** [review]
Thanks for your contribution @DerOetzi 
But I see some problems with this because it will break the query part of the addon:

- A complete solution should add also itemName as a tag when this option is used and make query part intelligent to use that logic.
- Also I would like to include test covering that new functionality.
- I added also some minor notes in your code

**Review by lujop — CHANGES_REQUESTED** [review]
Changes explained in previous comment

File: bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml
Comment (DerOetzi) [reply bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml]: Renamed

**Comment by DerOetzi** [comment]
> Thanks for your contribution @DerOetzi
> But I see some problems with this because it will break the query part of the addon:
> 
> * A complete solution should add also itemName as a tag when this option is used and make query part intelligent to use that logic.

The itemname is already always included as tag, if I see this correct. But your right I missed the query part. Please review my changes on this.

> * Also I would like to include test covering that new functionality.

I have added some tests. Please give me feedback, whether they are enough like this.

> * I added also some minor notes in your code

Renamed the things you commented

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/ItemToStorePointCreator.java
Comment (Hilbrand) [inline bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/ItemToStorePointCreator.java:79]: This code to search te MetadataRegistry is 3 times the same. Can you put it in a method that used at all these 3 locations?

**Review by Hilbrand — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/ItemToStorePointCreator.java
Comment (DerOetzi) [reply bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/ItemToStorePointCreator.java]: Refactored to a new utils class

File: bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml
Comment (DerOetzi) [reply bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml]: @lujop Can you have a look on the new description?

**Review by Hilbrand — APPROVED** [review]
LGTM. @lujop Can you have a second look at this pull request and approve if ok.

File: bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml
Comment (lujop) [inline bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml:84]: Add that parameter also to the documentation in the README. 
And a reference where it says ` with names equal to the...`
Also, it's important to document in the README the name of the metadata key the user must use.

**Review by lujop — CHANGES_REQUESTED** [review]
Thank you for your modifications, I haven't had time to test but it seems ok.
Only two minor things.

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBMetadataUtils.java
Comment (DerOetzi) [reply bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBMetadataUtils.java:40]: Okay will change this!

File: bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml
Comment (DerOetzi) [reply bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml:84]: Will add this and all other Metadata configuration to README as a complete new section.

> And a reference where it says with names equal to the...

I don't understand what you want me to do? Can you please explain?

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBMetadataUtils.java
Comment (DerOetzi) [reply bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBMetadataUtils.java:40]: While writing documentation about usage of metadata for additional information, I wonder now whether this configuration option is needed at all. User will configure it's item like this:

```
Number testItem1 {... influxdb= "" [tag1=foo, tag2= bar]}
or
Number testItem2 {... influxdb= "measurement_name" [tag1=foo, tag2=bar]}
```
Thinking again about this, it could be handled without this additional configuration? What do you think @lujop and @Hilbrand?

**Comment by DerOetzi** [comment]
I did some refactoring of my code as mentioned above there is no configuration flag for service anymore, but it just checks whether there is a metadata value or not. 

But while doing the refactoring I think I have found a bug inside the existing code, of ItemToStorePointCreator which wasn't noticed because of a missing test case. But I'm not sure and need your experience with the code @lujop and @Hilbrand.

At the `InfluxFilterCriteriaQueryCreatorImplTest.java`  there is a testcase `testEscapeSimpleItem` which test an item with a dot in its name. This one fails now with an IllegalArgumentException when trying to generate MetadataKey from item name.

I have added now such testcase to the ItemToStorePointCreator file as well, which fails although with the old code, without my changes because it tries to access MetadataKey in addPointTags as well.

My question is it a valid item name with a dot, then the question is how to access metadata with such items, or is this testcase invalid in the FilterQuery test?

File: bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml
Comment (lujop) [reply bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml:84]: In line 9 of the README:
Change:

    - The states of an item are persisted in _measurements_ points with names equal to the name of the item, or the alias, if one is provided. In both variants, a _tag_ named "item" is added, containing the item name.

for
  - The states of an item are persisted in _measurements_ points with names equal to the name of the item, its alias, or from some metadata depending on the configuration. In all variants, a _tag_ named "item" is added, containing the item name.

**Comment by DerOetzi** [comment]
I did some further research from what I see so far an item name cannot contain a dot because this will always fail here:

https://github.com/openhab/openhab-core/blob/029280683087d07d3d52bc1b125e73c17a33917a/bundles/org.openhab.core/src/main/java/org/openhab/core/common/AbstractUID.java#L96

So I think the testcase is not valid and I would remove it?

File: bundles/org.openhab.persistence.influxdb/src/test/java/org/openhab/persistence/influxdb/internal/ItemToStorePointCreatorTest.java
Comment (DerOetzi) [inline bundles/org.openhab.persistence.influxdb/src/test/java/org/openhab/persistence/influxdb/internal/ItemToStorePointCreatorTest.java:83]: As described in my comment above this test will fail although with the old code, because I think a item name with a dot inside is not valid

https://github.com/openhab/openhab-core/blob/029280683087d07d3d52bc1b125e73c17a33917a/bundles/org.openhab.core/src/main/java/org/openhab/core/common/AbstractUID.java#L96

But in filter criteria Test there is an existing testcase which fails now as well.

**Comment by DerOetzi** [comment]
@lujop and @Hilbrand removed the testcase for item name with dot

File: bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml
Comment (DerOetzi) [reply bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml:84]: Changed

File: bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml
Comment (DerOetzi) [reply bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml]: Isn't relevant anymore!

File: bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx1/Influx1FilterCriteriaQueryCreatorImpl.java
Comment (lujop) [inline bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx1/Influx1FilterCriteriaQueryCreatorImpl.java:64]:
```suggestion
        if (itemName != null && !tableName.equals(itemName))) {
```

**Review by lujop — APPROVED** [review]
First of all, I apologize for not being able to review the addon before.
Today I've been able to build and test it for InfluxDB 2 and worked correctly.

The code seems good to me also. 
I only made a little commit with some minor documentation changes and suggested a change in a condition expanding the negation to make  reading the condition clearer.

Thanks for the PR

File: bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml
Comment (lujop) [inline bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml:91]:
```suggestion
```
In the review I forgot to say that his must be removed as it isn't any longer used.

**Review by lujop — CHANGES_REQUESTED** [review]
Just remove the unused parameter from config.

**Comment by DerOetzi** [comment]
> First of all, I apologize for not being able to review the addon before.
> Today I've been able to build and test it for InfluxDB 2 and worked correctly.
> 
> The code seems good to me also.
> I only made a little commit with some minor documentation changes and suggested a change in a condition expanding the negation to make reading the condition clearer. Please accept the commit suggestion.
> 
> Good work, and thank you very much for the PR.

Thanks for your constructive review comments. Think we found a good solution together now.

**Comment by DerOetzi** [comment]
Please don't merge yet I think I have found another problem on restoring item states

**Comment by DerOetzi** [comment]
> Please don't merge yet I think I have found another problem on restoring item states
 
@lujop can you please recheck my last changes to V1 Query parts. As I noticed there was a bug because tem-tags where not accquired.

**Comment by lujop** [comment]
> > Please don't merge yet I think I have found another problem on restoring item states
> 
> @lujop can you please recheck my last changes to V1 Query parts. As I noticed there was a bug because tem-tags where not accquired.

Sure, let me find some time and I will install InfluxDB1 and do some more detailed testing

**Comment by DerOetzi** [comment]
> > > Please don't merge yet I think I have found another problem on restoring item states
> > 
> > 
> > @lujop can you please recheck my last changes to V1 Query parts. As I noticed there was a bug because tem-tags where not accquired.
> 
> Sure, let me find some time and I will install InfluxDB1 and do some more detailed testing

I have installed a v1.8 docker container and did some more testing myself. For me everything seems to work fine now, but I would be very glad, if find the time to check it as well @lujop

**Review by lujop — APPROVED** [review]
I've done some testing with InfluxDB1 and it seems to be working correctly.
Thanks for the fixs.

**Comment by lujop** [comment]
@Hilbrand It seems ok to be integrated for me.

**Comment by DerOetzi** [comment]
@Hilbrand and @lujop is there any problem with this PR? Why isn't it merged yet?

**Comment by kaikreuzer** [comment]
As @Hilbrand already approved on Feb 3, I'd say we can merge. 👍

**Pull request merged**
